huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go

```diff
@@ -0,0 +1,418 @@
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(objectId)
+
+	return resourceEipProtectionRead(ctx, d, meta)
+}
+
+func buildModifyProtectedEipsParams(objectId string, protectedEips *schema.Set,
+	operation int) (map[string]interface{}, []string) {
+	protectEipInfo, protectedEipIds := buildProtectedEipBodyAndIds(protectedEips)
+	requestParams := map[string]interface{}{
+		"object_id": objectId,
+		"ip_infos":  protectEipInfo,
+		"status":    operation,
+	}
+	return requestParams, protectedEipIds
+}
+
+func buildProtectedEipBodyAndIds(rawParams *schema.Set) ([]map[string]interface{}, []string) {
+	if rawParams.Len() < 1 {
+		return nil, nil
+	}
+
+	requestParams := make([]map[string]interface{}, rawParams.Len())
+	protectedEipIds := make([]string, rawParams.Len())
+	for i, v := range rawParams.List() {
+		raw := v.(map[string]interface{})
+		requestParams[i] = map[string]interface{}{
+			"id":          raw["id"],
+			"public_ip":   utils.ValueIngoreEmpty(raw["public_ipv4"]),
+			"public_ipv6": utils.ValueIngoreEmpty(raw["public_ipv6"]),
+		}
+		protectedEipIds[i] = raw["id"].(string)
+	}
+	return requestParams, protectedEipIds
+}
+
+// ProtectedEipExist method will return true if a protected public IP exists under the object.
+func ProtectedEipExist(eips []interface{}) bool {
+	for _, eip := range eips {
+		status := utils.PathSearch("status", eip, 1)
+		if int(status.(float64)) == openEipProtection {
+			return true
+		}
+	}
+	return false
+}
+
+func resourceEipProtectionRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg      = meta.(*config.Config)
+		region   = cfg.GetRegion(d)
+		objectId = d.Id()
+	)
+	client, err := cfg.NewServiceClient(eipProtectionProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating CFW Client: %s", err)
+	}
+
+	resp, err := QuerySyncedEips(client, queryHttpUrl, objectId)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if !ProtectedEipExist(resp) {
+		return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{}, "No EIP is being protected")
+	}
+
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("protected_eip", flattenProtectedEips(resp)),
+	)
+	if err = mErr.ErrorOrNil(); err != nil {
+		return diag.Errorf("error saving EIP protection resource fields: %s", err)
+	}
+	return nil
+}
+
+func flattenProtectedEips(eips []interface{}) []interface{} {
+	rst := make([]interface{}, 0, len(eips))
+	for _, eip := range eips {
+		if status := utils.PathSearch("status", eip, 1); int(status.(float64)) != openEipProtection {
+			continue
+		}
+		rst = append(rst, map[string]interface{}{
+			"id":          utils.PathSearch("id", eip, nil),
+			"public_ipv4": utils.PathSearch("public_ip", eip, nil),
+			"public_ipv6": utils.PathSearch("public_ipv6", eip, nil),
+		})
+	}
+	return rst
+}
+
+func unprotectEips(ctx context.Context, client *golangsdk.ServiceClient, objectId string, unprotectedEipSet *schema.Set,
```

Comment:
it's better to re-arrange some functions, for example: put protectEips and unprotectEips together, put exportable functions together

Reply:
done.